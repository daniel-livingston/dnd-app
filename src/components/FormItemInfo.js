import React from "react";
import PropTypes from "prop-types";

const propTypes = {
	title: PropTypes.string.isRequired,
	description: PropTypes.string,
};

const FormItemInfo = (props) => (
	<div className='FormItemInfo'>
		<h3 className='FormItemInfo-Title'>{props.title}</h3>
		{props.description && <p className='FormItemInfo-Description'>{props.description}</p>}
	</div>
);

FormItemInfo.propTypes = propTypes;

export default FormItemInfo;
