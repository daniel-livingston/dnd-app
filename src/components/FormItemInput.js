import React from "react";
import PropTypes from "prop-types";

const propTypes = {
	type: PropTypes.string.isRequired,
	value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
	onChange: PropTypes.func.isRequired,
	placeholder: PropTypes.string.isRequired,
	options: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),
};

const FormItemInput = (props) => {
	const getInput = (type, onChange, placeholder, value, options) => {
		switch (type) {
			case "number": {
				return (
					<input
						className='FormItemInput FormItemInput-Number'
						onChange={props.onChange}
						type='number'
						value={props.value}
					/>
				);
			}
			case "password": {
				return (
					<input
						className='FormItemInput FormItemInput-Password'
						onChange={onChange}
						placeholder={placeholder}
						type='text'
						value={value}
					/>
				);
			}
			case "select": {
				if (!options) throw Error("No options array provided");
				return (
					<select
						className='FormItemInput FormItemInput-Select'
						onChange={onChange}
						value={value}
					>
						{options.map((option) => (
							<option key={option} value={option}>
								{option}
							</option>
						))}
					</select>
				);
			}
			case "text": {
				return (
					<input
						className='FormItemInput FormItemInput-Text'
						onChange={onChange}
						placeholder={placeholder}
						type='text'
						value={value}
					/>
				);
			}
		}
	};

	return getInput(props.type, props.onChange, props.placeholder, props.value, props.options);
};

FormItemInput.propTypes = propTypes;

export default FormItemInput;
