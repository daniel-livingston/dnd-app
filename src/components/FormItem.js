import React from "react";
import PropTypes from "prop-types";
import FormItemInfo from "./FormItemInfo";
import FormItemInput from "./FormItemInput";

const propTypes = {
	title: PropTypes.string.isRequired,
	description: PropTypes.string,
	inputType: PropTypes.string.isRequired,
	inputValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
	inputOnChange: PropTypes.func.isRequired,
	inputPlaceholder: PropTypes.string.isRequired,
	selectOptions: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),
};

const FormItem = (props) => (
	<div className='FormItem'>
		<FormItemInfo title={props.title} description={props.description} />
		<FormItemInput
			type={props.inputType}
			value={props.inputValue}
			onChange={props.inputOnChange}
			placeholder={props.inputPlaceholder}
			options={props.selectOptions}
		/>
	</div>
);

FormItem.propTypes = propTypes;

export default FormItem;
