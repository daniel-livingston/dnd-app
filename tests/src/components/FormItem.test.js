import React from "react";
import { shallow } from "enzyme";
import FormItem from "../../../src/components/FormItem";
import FormItemInfo from "../../../src/components/FormItemInfo";
import FormItemInput from "../../../src/components/FormItemInput";

const formItemInfoProps = {
	title: "Test title",
	description: "Test description",
};

const formItemInputProps = {
	inputType: "text",
	inputValue: "Test value",
	inputOnChange: jest.fn(),
	inputPlaceholder: "Test placeholder",
};

describe("<FormItem>", () => {
	test("renders correctly", () => {
		const wrapper = shallow(<FormItem {...formItemInfoProps} {...formItemInputProps} />);
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find(FormItemInfo).props()).toEqual(formItemInfoProps);
		expect(wrapper.find(FormItemInput).props()).toEqual({
			type: formItemInputProps.inputType,
			value: formItemInputProps.inputValue,
			onChange: formItemInputProps.inputOnChange,
			placeholder: formItemInputProps.inputPlaceholder,
			options: undefined,
		});
	});
});
