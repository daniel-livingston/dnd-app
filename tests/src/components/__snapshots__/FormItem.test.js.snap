// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormItem> renders correctly 1`] = `
<div
  className="FormItem"
>
  <FormItemInfo
    description="Test description"
    title="Test title"
  />
  <FormItemInput
    onChange={[MockFunction]}
    placeholder="Test placeholder"
    type="text"
    value="Test value"
  />
</div>
`;
